### PR TITLE
improve logits bias

### PR DESCRIPTION
**Pull request by yuguo68**
Summary:
Signed-off-by: Yu Guo <yuguo@meta.com>

use `torch.index_put_(., accumulate=True)` to update the logits tensor once.

Rollback Plan:

pytest -s -v tests/v1/sample/test_sampler.py

Differential Revision: D75833790

### CODE REVIEW

File: vllm/v1/sample/sampler.py [hunk 1]
```diff
@@ -243,7 +263,15 @@ def apply_logits_bias(
...
                             f"out-of-vocab token id. Vocabulary size: "
                             f"{vocab_size}")
-                    logits[i, token_id] += bias
+                        rows += [i]
+                        cols += [token_id]
+                        vals += [bias]
+
+        rows = list_to_gpu(rows, self.pin_memory, logits.device)
+        cols = list_to_gpu(cols, self.pin_memory, logits.device)
+        values = list_to_gpu(vals, self.pin_memory, logits.device)
```
Comment (ngimel) [inline vllm/v1/sample/sampler.py]: isn't bias on device already?

Comment (yuguo68) [reply vllm/v1/sample/sampler.py]: it is a python float, https://github.com/vllm-project/vllm/blob/7594173c9e05aedb82e92ded0c1b04b2a7c684f2/vllm/v1/sample/metadata.py#L36

**Review by houseroad — COMMENTED** [review]
Add some test plan about how to verify the correctness?

File: vllm/v1/sample/sampler.py [hunk 2]
```diff
@@ -243,7 +263,13 @@ def apply_logits_bias(
                             f"token_id {token_id} in logit_bias contains "
                             f"out-of-vocab token id. Vocabulary size: "
                             f"{vocab_size}")
-                    logits[i, token_id] += bias
+                        rows.append(i)
+                        cols.append(token_id)
+                        vals.append(bias)
+
+        indices = list_to_gpu([rows, cols], self.pin_memory, logits.device)
+        values = list_to_gpu(vals, self.pin_memory, logits.device)
```
Comment (houseroad) [inline vllm/v1/sample/sampler.py]: please fix the linter :-)

**Comment by njhill** [comment]
Thanks @yuguo68. FYI we are planning to replace most of the sampling param implementations with vectorized versions soon, see https://github.com/vllm-project/vllm/pull/13360/files#diff-d01f143e1af472f24af24842cb879907ce624e6e5c977935e944545240723529R122

File: vllm/v1/sample/sampler.py (hunk 2 above)
Comment (yuguo68) [reply vllm/v1/sample/sampler.py]: updated list to Sequence for the linter, checked https://github.com/python/mypy/issues/3351 for my own reference.

**Comment by yuguo68** [comment]
> Add some test plan about how to verify the correctness?

I think the unittest https://github.com/vllm-project/vllm/blob/bdce64f2365b39335141f8efcb3a0a8ecc559153/tests/v1/sample/test_sampler.py#L446 should cover it. let me double check.

**Comment by yuguo68** [comment]
> Thanks @yuguo68. FYI we are planning to replace most of the sampling param implementations with vectorized versions soon, see https://github.com/vllm-project/vllm/pull/13360/files#diff-d01f143e1af472f24af24842cb879907ce624e6e5c977935e944545240723529R122

@njhill oh, thanks for the link! The implementation is actually the same except I just learnt from @Adolfo-Karim we can create a single tensor for the indices instead of two. Please let me know whether we should proceed with this PR.

**Comment by houseroad** [comment]
@yuguo68 , if we have obvious gain, we can still land the PR first, since landing the official new design still need some more time.

**Comment by yuguo68** [comment]
> @yuguo68 , if we have obvious gain, we can still land the PR first, since landing the official new design still need some more time.

sounds good, I will do a microbenchmark.

**Comment by yuguo68** [comment]
microbenchmark with script on H100

```

from typing import Optional, Union

import torch

def async_tensor_h2d(
    data: list,
    dtype: torch.dtype,
    target_device: Union[str, torch.device],
    pin_memory: bool,
) -> torch.Tensor:
    """Asynchronously create a tensor and copy it from host to device."""
    t = torch.tensor(data, dtype=dtype, pin_memory=pin_memory, device="cpu")
    return t.to(device=target_device, non_blocking=True)


def apply_logits_bias_opt(
    logits: torch.Tensor,
    logit_bias_list: list[Optional[dict[int, float]]],
) -> torch.Tensor:

    rows: list[int] = []
    cols: list[int] = []
    vals: list[float] = []

    # Get vocabulary size from logits
    vocab_size = logits.shape[-1]

    for i, logit_bias in enumerate(logit_bias_list):
        if logit_bias:
            for token_id, bias in logit_bias.items():
                # Check token_id bounds to ensure within vocabulary
                if token_id < 0 or token_id >= vocab_size:
                    raise ValueError(
                        f"token_id {token_id} in logit_bias contains "
                        f"out-of-vocab token id. Vocabulary size: "
                        f"{vocab_size}"
                    )
                rows.append(i)
                cols.append(token_id)
                vals.append(bias)

    if rows:
        indices = async_tensor_h2d([rows, cols], torch.int64,
                                    logits.device, True)
        values = async_tensor_h2d(vals, torch.float, logits.device,
                                    True)
        logits.index_put_(tuple(indices), values=values, accumulate=True)

    torch.cuda.synchronize()
    return logits


def apply_logits_bias(
    logits: torch.Tensor,
    logit_bias_list: list[Optional[dict[int, float]]],
) -> torch.Tensor:

    # Get vocabulary size from logits
    vocab_size = logits.shape[-1]

    for i, logit_bias in enumerate(logit_bias_list):
        if logit_bias:
            for token_id, bias in logit_bias.items():
                # Check token_id bounds to ensure within vocabulary
                if token_id < 0 or token_id >= vocab_size:
                    raise ValueError(
                        f"token_id {token_id} in logit_bias contains "
                        f"out-of-vocab token id. Vocabulary size: "
                        f"{vocab_size}"
                    )
                logits[i, token_id] += bias

    torch.cuda.synchronize()
    return logits


mbs = 8
vocab_size = 32000

import random

logits = torch.rand(mbs, vocab_size, device="cuda")

bs = 5

for num_bias in [10, 50, 100]:
    logit_bias_list = []
    for _ in range(bs):
        logit_bias = {}
        for _ in range(num_bias):
            logit_bias[random.randint(0, vocab_size - 1)] = random.random()
        logit_bias_list.append(logit_bias)
    assert torch.allclose(
        apply_logits_bias_opt(logits, logit_bias_list),
        apply_logits_bias(logits, logit_bias_list),
    )
    print(f"---------num_bias: {num_bias}, batch size: {bs}---------")
    print("this PR: batched logits update")
    %timeit apply_logits_bias_opt(logits, logit_bias_list)
    print("trunk non-batched logits update")
    %timeit apply_logits_bias(logits, logit_bias_list)    
```

Output:
```
---------num_bias: 10, batch size: 5---------
this PR: batched logits update
156 µs ± 884 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
trunk non-batched logits update
368 µs ± 562 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
---------num_bias: 50, batch size: 5---------
this PR: batched logits update
229 µs ± 3.37 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
trunk non-batched logits update
1.8 ms ± 6.08 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
---------num_bias: 100, batch size: 5---------
this PR: batched logits update
309 µs ± 370 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
trunk non-batched logits update
3.62 ms ± 4.67 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by yuguo68** [comment]
@houseroad , @njhill added micro benchmark, we do have significant improvement with batched update especially when the number of updates is large, e.g., 3.62 ms to 309 µs with 500 updates.

**Review by houseroad — APPROVED** [review]
Pretty nice improvement. Thanks. 

For the benchmark, we can consider to upstream it here: https://github.com/vllm-project/vllm/tree/main/benchmarks (it can be a follow up task)
